copy any file from stage0/lib to stage0-sysroot/lib

onur-ozkan opened the pull request:
With the LLVM 18 upgrade, the name of the LLVM library has been changed to something like `libLLVM.so.18.1-rust-1.78.0-beta`, which `is_dylib` function cannot determine as it only looks whether files are ending with ".so" or not.
This change resolves this problem by no longer doing that ".so" check, as we need all files from the stage0/lib as they are all dependency of rustc anyway.

Fixes #122913

onur-ozkan committed:
copy any file from stage0/lib to stage0-sysroot/lib

With the LLVM 18 upgrade, the name of the LLVM library has been changed to something like
`libLLVM.so.18.1-rust-1.78.0-beta`, which `is_dylib` function cannot determine as it only
looks whether files are ending with ".so" or not.
This change resolves this problem by no longer doing that ".so" check, as we need all files
from the stage0/lib as they are all dependency of rustc anyway.

Signed-off-by: onur-ozkan <work@onurozkan.dev>

Kobzol commented:
Confirmed that the LLVM file is copied into `stage0-sysroot` and `cargo +stage0 build` works.

@bors r+ rollup

bors commented:
:pushpin: Commit 5fe364afdd590c8f928721820e490f82b90f65e0 has been approved by `Kobzol`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Kobzol 5fe364afdd590c8f928721820e490f82b90f65e0 -->
<!-- homu: {"type":"Approved","sha":"5fe364afdd590c8f928721820e490f82b90f65e0","approver":"Kobzol","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

Nilstrieb commented:
I thought I fixed is_dylib when I fixed the nix patching?

Kobzol commented:
That might be an orthogonal issue, but I would personally prefer to copy everything anyway, we just want to duplicate the `lib` directory to build the stage0 sysroot, so it seems more forward compatible to copy everything.

onur-ozkan commented:
> I thought I fixed is_dylib when I fixed the nix patching?

Library name is `libLLVM.so.18.1-rust-1.78.0-beta`, `is_lib` doesn't do anything useful to detect it. https://github.com/rust-lang/rust/blob/760e567af5398a0d8c512f904e551e1f38e00d79/src/bootstrap/src/utils/helpers.rs#L56-L59

onur-ozkan commented:
We also have a linker script file that ends with `.so` which isn't a dynamic library anyway.

Nilstrieb commented:
ah no, I created a competing implementation https://github.com/rust-lang/rust/pull/122911/files#diff-11a177c4925352bc8852241c59fd49c369df531495f11f81914eb6972783bb8aR747
